### PR TITLE
Deploy rust-code-analysis for Windows

### LINKED ISSUE #137: Deploy rust-code-analysis for Windows
See https://github.com/mozilla/rust-code-analysis/blob/master/.taskcluster.yml#L89

### PR DESCRIPTION
**Pull request by Luni-4**
This PR fixes #137 

Thanks in advance for your review! :)

### CODE REVIEW

File: .taskcluster.yml [hunk 1]
```diff
@@ -138,17 +138,61 @@ tasks:
...
                  git -c advice.detachedHead=false checkout ${head_rev} &&
-                 cargo build --release &&
+                 cargo build --all --release &&
                  cd target/release &&
-                 tar -zvcf /build.tar.gz --transform 's,^,rust-code-analysis-linux-x86_64/,' rust-code-analysis"
+                 mkdir /build.tar.gz &&
+                 mv rust-code-analysis-cli /build.tar.gz"
             artifacts:
-              public/rust-code-analysis-linux-x86_64.tar.gz:
+              public/rust-code-analysis-cli:
```
Comment (marco-c) [inline .taskcluster.yml:146]: This is the path in the URL

File: .taskcluster.yml [hunk 2]
```diff
@@ -138,17 +138,61 @@ tasks:
...
                 expires: {$fromNow: '2 weeks'}
                 path: /build.tar.gz
```
Comment (marco-c) [inline .taskcluster.yml:148]: And this is the path in the Docker container

File: .taskcluster.yml (hunk 1 above)
Comment (Luni-4) [reply .taskcluster.yml:146]: I see, thanks!

File: .taskcluster.yml (hunk 2 above)
Comment (Luni-4) [reply .taskcluster.yml:148]: Is the path where I can find the `rust-code-analysis-cli` binary, right? So is it correct in this case?

Comment (marco-c) [reply .taskcluster.yml:148]: Yep! Although it's super ugly to call the binary "build.tar.gz" :P

Comment (Luni-4) [reply .taskcluster.yml:148]: You're right, I'm going to change that name :)

File: .taskcluster.yml [hunk 3]
```diff
@@ -142,17 +142,67 @@ tasks:
...
+              - FOR /F "tokens=*" %%o IN ('%VSWHERE_CMD%') do (SET LIB_PATH=%%o)
+              - set PATH=%LIB_PATH%;%PATH%
+              - rustup-init -yv --default-toolchain stable ^
+                                --default-host x86_64-pc-windows-msvc
+              - git clone --recursive --quiet ${repository}
+              - cd rust-code-analysis
+              - git -c advice.detachedHead=false checkout ${head_rev}
+              - cargo build --all --release
+              - cd target/release
+              - move rust-code-analysis-cli.exe C:\Users
```
Comment (Luni-4) [inline .taskcluster.yml:193]: I don't know if I can run this command without admin privileges. I don't know which folder should be used to avoid this inconvenient.

File: .taskcluster.yml [hunk 4]
```diff
@@ -178,11 +126,22 @@ tasks:
...
+                - cd /tmp
+                - strip $ARTIFACT_TWO
+                - zip $ARTIFACT_TWO_ZIP $ARTIFACT_TWO
                 - taskboot
                 - github-release
                 - mozilla/rust-code-analysis
                 - "${version}"
-                - "--asset=rust-code-analysis-linux-x86_64.tar.gz:public/rust-code-analysis-linux-x86_64.tar.gz"
+                - "--asset=$ARTIFACT_ONE:public/$ARTIFACT_ONE"
+                - "--asset=$ARTIFACT_TWO_ZIP:/tmp/$ARTIFACT_TWO_ZIP"
```
Comment (marco-c) [inline .taskcluster.yml:144]: I think `taskboot github-release` will try to take the assets from a dependency, we might have to add a `--local-asset` argument to support this.

File: .taskcluster.yml [hunk 5]
```diff
@@ -178,11 +126,22 @@ tasks:
...
+                - ARTIFACT_ONE=rust-code-analysis-linux-x86_64.tar.gz
+                - ARTIFACT_TWO=rust-code-analysis-cli.exe
+                - ARTIFACT_TWO_ZIP=rust-code-analysis-windows-x86_64.zip
+                - taskboot
+                - retrieve-artifact
+                - "--output-path=/tmp"
+                - "--artifacts=$ARTIFACT_TWO:public/$ARTIFACT_TWO"
+                - cd /tmp
+                - strip $ARTIFACT_TWO
+                - zip $ARTIFACT_TWO_ZIP $ARTIFACT_TWO
```
Comment (marco-c) [inline .taskcluster.yml:138]: Use the highest compression ratio: `zip -9`

File: .taskcluster.yml (hunk 4 above)
Comment (Luni-4) [reply .taskcluster.yml:144]: I guess we could strip in an intermediate task and then upload from there using the `--asset` for now. What do you think?

File: .taskcluster.yml [hunk 6]
```diff
@@ -142,17 +44,96 @@ tasks:
...
+          payload:
+            maxRunTime: 3600
+            image: "rust:buster"
+            env:
+              PROJECT_NAME: rust-code-analysis
+            command:
+                - WINDOWS_BINARY=rust-code-analysis-cli.exe
+                - taskboot
+                - retrieve-artifact
+                - "--output-path=/tmp"
```
Comment (marco-c) [inline .taskcluster.yml:124]: Nit: you can use `--output-path=.` or `--output-path=$PWD` so you don't have to `cd /tmp` later

File: .taskcluster.yml [hunk 7]
```diff
@@ -142,17 +44,96 @@ tasks:
...
+                - "--artifacts=$WINDOWS_BINARY:public/$WINDOWS_BINARY"
+                - cd /tmp
+                - strip $WINDOWS_BINARY
+                - zip -9 rust-code-analysis-windows-x86_64.zip WINDOWS_BINARY
+            artifacts:
+              public/rust-code-analysis-windows-x86_64.zip:
+                expires: {$fromNow: '2 weeks'}
+                path: /tmp
```
Comment (marco-c) [inline .taskcluster.yml:132]: This should be the path to the file, not to the directory containing the file.

File: .taskcluster.yml [hunk 8]
```diff
@@ -178,11 +160,14 @@ tasks:
               env:
                 TASKCLUSTER_SECRET: project/relman/rust-code-analysis/deploy
               command:
+                - LINUX_ARTIFACT=rust-code-analysis-linux-x86_64.tar.gz
+                - WINDOWS_ARTIFACT=rust-code-analysis-windows-x86_64.zip
```
Comment (marco-c) [inline .taskcluster.yml:164]: Define them in the `env` section right above

File: .taskcluster.yml [hunk 9]
```diff
@@ -142,17 +44,96 @@ tasks:
...
+          deadline: {$fromNow: '2 hour'}
+          provisionerId: proj-relman
+          workerType: ci
+          payload:
+            maxRunTime: 3600
+            image: "rust:buster"
+            env:
+              PROJECT_NAME: rust-code-analysis
+            command:
+                - WINDOWS_BINARY=rust-code-analysis-cli.exe
```
Comment (marco-c) [inline .taskcluster.yml:121]: Define it in the `env` section right above

File: .taskcluster.yml [hunk 10]
```diff
@@ -142,17 +44,96 @@ tasks:
...
+        then:
+          taskId: {$eval: as_slugid("build_windows_release")}
+          created: {$fromNow: ''}
+          deadline: {$fromNow: '2 hour'}
+          provisionerId: proj-relman
+          workerType: win2012r2
+          payload:
+            maxRunTime: 3600
+            env:
+              PROJECT_NAME: rust-code-analysis
```
Comment (marco-c) [inline .taskcluster.yml:74]: Why is this needed?

File: .taskcluster.yml [hunk 11]
```diff
@@ -26,133 +26,108 @@ tasks:
...
+              - set VSWHERE_CMD=vswhere -latest -products * -find %VS_BINS%
+              - FOR /F "tokens=*" %%o IN ('%VSWHERE_CMD%') do (SET LIB_PATH=%%o)
+              - set PATH=%LIB_PATH%;%PATH%
+              - rustup-init -yv --default-toolchain stable ^
+                                --default-host x86_64-pc-windows-msvc
+              - git clone --recursive --quiet ${repository}
+              - cd rust-code-analysis
+              - git -c advice.detachedHead=false checkout ${head_rev}
+              - cargo build --all --release
+              - move target\release\rust-code-analysis-cli.exe target
```
Comment (marco-c) [inline .taskcluster.yml:86]: No need to move the file, just set the artifact's `path` to the right path (so it would be `rust-code-analysis/target/release/rust-code-analysis-cli.exe` instead of `rust-code-analysis/target/rust-code-analysis-cli.exe`)

**Comment by marco-c** [comment]
LGTM! Can you apply the last suggested change and revert the removals?

**Comment by marco-c** [comment]
Can you rebase on top of master? There was an issue in the .taskcluster.yml linter that I just fixed.

**Comment by marco-c** [comment]
(not sure why I can usually rebase PRs myself via GitHub's UI, but I can't with your PRs. Maybe you need to do this: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
